File: axi/src/axi/sync/stamping.py
"""The Lamport clock and the origin stamp: making LWW real instead of aspirational.

`nodes.lamport` and `nodes.origin_node` have existed since the sync-schema
chain and NOTHING has ever written them. Every row in every LifeOS database
carries `lamport = 0` and `origin_node = NULL`. A merge engine fed rows that
all claim clock 0 cannot order anything — it would resolve every real conflict
by whatever happened to arrive last, which is precisely the silent data loss
the design exists to prevent.

This module starts the clock, stamps every write, and backfills the rows that
were written before any of it existed.

ON THE BACKFILL BEING TRUTHFUL. Pre-existing rows keep `lamport = 0`. Inventing
a plausible-looking value for them would be a lie the merge engine then trusts;
0 says "older than anything written since the clock started", which is exactly
what they are. What they DO get is an origin, because they were authored here.
"""

from __future__ import annotations

import threading
from typing import Any

from axi import store

#: Where this device's identity is cached. Read once per process: a Lamport
#: stamp on every insert must not cost a query.
_ORIGIN_LOCK = threading.Lock()
_ORIGIN: str | None = None

#: The monotonic counter. Guarded because axi writes from several threads (the
#: daemon, the dashboard, background jobs) and two writes sharing a Lamport
#: value would make them unorderable — the tiebreak would then decide conflicts
#: it was never meant to see.
_CLOCK_LOCK = threading.Lock()
_CLOCK: int | None = None


class ParityLost(RuntimeError):
    """The migration's before/after row counts disagree. Nothing may proceed."""


def local_origin(conn=None) -> str:
    """This device's UUID, as written into `origin_node`.

    Created on first use and stored in `meta`, so it survives restarts: a
    device whose origin changed would look like a different author for
    everything it wrote afterwards, and its own rows would start conflicting
    with each other.
    """
    global _ORIGIN
    with _ORIGIN_LOCK:
        if _ORIGIN is not None:
            return _ORIGIN

        c = conn or store._connect()  # noqa: SLF001
        row = c.execute(
            "SELECT value FROM meta WHERE key = 'sync_origin_node'"
        ).fetchone()
        if row is None:
            from axi.sync import identity

            origin = identity.new_device_uuid()
            c.execute(
                "INSERT OR REPLACE INTO meta(key, value) VALUES ('sync_origin_node', ?)",
                (origin,),
            )
            c.commit()
        else:
            origin = row["value"]

        _ORIGIN = origin
        return origin


def reset_cache() -> None:
    """Drop the cached origin and clock. For tests that swap databases."""
    global _ORIGIN, _CLOCK
    with _ORIGIN_LOCK:
        _ORIGIN = None
    with _CLOCK_LOCK:
        _CLOCK = None


def next_lamport(conn=None) -> int:
    """The next clock value for a local write. Strictly increasing, never repeated."""
    global _CLOCK
    with _CLOCK_LOCK:
        if _CLOCK is None:
            c = conn or store._connect()  # noqa: SLF001
            highest = 0
            for table in ("nodes", "edges"):
                row = c.execute(f"SELECT MAX(lamport) AS m FROM {table}").fetchone()
                if row and row["m"] is not None:
                    highest = max(highest, int(row["m"]))
            _CLOCK = highest
        _CLOCK += 1
        return _CLOCK


def observe_lamport(value: int) -> None:
    """Merge a peer's clock into ours.

    Lamport's rule: on receiving an event, the local clock becomes
    `max(local, received)`. Without this, a device that has been quiet would
    keep stamping low values and its genuinely newer edits would lose every
    conflict against a chattier peer.
    """
    global _CLOCK
    with _CLOCK_LOCK:
        if _CLOCK is None or value > _CLOCK:
            _CLOCK = value


# --------------------------------------------------------------------------
# migration
# --------------------------------------------------------------------------

_PARITY_TABLES = ("nodes", "edges")


def parity_reference(conn) -> dict[str, Any]:
    """Snapshot what the database holds BEFORE the migration touches it.

    Same discipline the schema-rebuild migration used: a migration that cannot
    state what it started with cannot prove it lost nothing.
    """
    reference: dict[str, Any] = {}
    for table in _PARITY_TABLES:
        row = conn.execute(
            f"SELECT COUNT(*) AS n, MAX(rowid) AS m FROM {table}"
        ).fetchone()
        reference[table] = {"count": row["n"], "max_rowid": row["m"]}
    return reference


def verify_parity(conn, reference: dict[str, Any]) -> None:
    """Raise [ParityLost] if any row went missing. Must be able to FAIL."""
    current = parity_reference(conn)
    for table, before in reference.items():
        after = current[table]
        if after["count"] < before["count"]:
            raise ParityLost(
                f"{table}: {before['count']} rows before the migration, "
                f"{after['count']} after — refusing to continue"
            )


def backfill(conn=None) -> dict[str, Any]:
    """Give pre-clock rows an origin, and start the counter above them.

    Idempotent: runs at every startup. Only rows with a NULL `origin_node` are
    touched, so a row synced FROM another device keeps that device as its
    author — overwriting it would make every device claim authorship of
    everything it ever received, and the deterministic tiebreak would stop
    meaning anything.
    """
    c = conn or store._connect()  # noqa: SLF001
    reference = parity_reference(c)
    origin = local_origin(c)

    for table in _PARITY_TABLES:
        c.execute(
            f"UPDATE {table} SET origin_node = ? WHERE origin_node IS NULL",
            (origin,),
        )
    c.commit()

    verify_parity(c, reference)

    # Reset so the counter re-reads MAX(lamport) including anything the
    # backfill saw; a stale cached counter would restart under existing rows.
    global _CLOCK
    with _CLOCK_LOCK:
        _CLOCK = None

    return reference


def ensure_sync_tables(conn=None) -> None:
    """Bookkeeping the sync engine needs, created alongside the backfill."""
    c = conn or store._connect()  # noqa: SLF001
    c.executescript(
        """
        -- How far we have got with each peer, so a push sends only what is new.
        CREATE TABLE IF NOT EXISTS sync_peer_state (
            peer_uuid   TEXT PRIMARY KEY,
            cursor      INTEGER NOT NULL DEFAULT -1,
            updated_at  REAL NOT NULL
        );

        -- Envelopes already applied. The relay guarantees at-least-once
        -- delivery, never exactly-once, so idempotency has to live here: the
        -- same envelope arriving twice must be a no-op, not a double apply.
        CREATE TABLE IF NOT EXISTS sync_applied (
            env_id     TEXT PRIMARY KEY,
            applied_at REAL NOT NULL
        );

        -- Every revision that LOST a merge. Created at startup, not on the
        -- first conflict: a table that springs into existence the moment
        -- something goes wrong is a table whose absence reads as "no conflicts"
        -- and whose queries fail on a healthy device.
        --
        -- The UNIQUE key is what stops a redelivered envelope from making the
        -- user stare at the same decision twice.
        CREATE TABLE IF NOT EXISTS sync_conflicts (
            id             INTEGER PRIMARY KEY AUTOINCREMENT,
            uuid           TEXT NOT NULL,
            losing_lamport INTEGER NOT NULL,
            losing_origin  TEXT,
            losing_payload TEXT NOT NULL,
            resolved_at    REAL NOT NULL,
            UNIQUE(uuid, losing_lamport, losing_origin, losing_payload)
        );
        """
    )
    c.commit()


def remember_applied(conn, env_id: str) -> bool:
    """Record an envelope as applied. False when it had already been seen."""
    import time

    ensure_sync_tables(conn)
    cur = conn.execute(
        "INSERT OR IGNORE INTO sync_applied(env_id, applied_at) VALUES (?, ?)",
        (env_id, time.time()),
    )
    conn.commit()
    return cur.rowcount == 1

File: mobile/lib/core/sync/stamping.dart
// Stamping local writes so they can ever be synced.
//
// The Dart mirror of `axi/src/axi/sync/stamping.py`. Without it the phone's
// graph writes rows with `lamport = 0` and a NULL origin — and every query the
// engine runs is `WHERE lamport > cursor`, so an unstamped row is invisible to
// sync for ever. It is not "sync is slow"; it is "nothing exists to send".
//
// Two values go on every write:
//
//   * `origin_node` — WHICH device authored it. Used only to break ties, but
//     without it two simultaneous edits have no deterministic winner and the
//     two devices can settle on DIFFERENT rows, which is worse than either
//     choice: they stop converging and nothing reports it.
//   * `lamport` — a logical clock, not a timestamp. Wall clocks on phones jump
//     (timezones, NTP, a user setting the date), and a clock that goes backwards
//     makes a new edit lose to an old one permanently.
import 'package:sqflite_common/sqlite_api.dart';
import 'package:uuid/uuid.dart';

const String kSyncIdentityTable = 'sync_identity';
const String kSyncPeerStateTable = 'sync_peer_state';
const String kSyncAppliedTable = 'sync_applied';
const String kSyncConflictsTable = 'sync_conflicts';

/// A peer we have never heard from. `-1` and not `0` because lamport values
/// start at 1: with `0` the very first row would fail `lamport > cursor` and
/// never ship.
const int kCursorUnsynced = -1;

const _uuid = Uuid();

/// Create the bookkeeping tables. Idempotent, and called at startup rather
/// than on first use: a table that springs into existence the moment something
/// goes wrong is a table whose absence reads as "no conflicts" and whose
/// queries fail on a perfectly healthy device.
Future<void> ensureSyncTables(DatabaseExecutor db) async {
  await db.execute('''
    CREATE TABLE IF NOT EXISTS $kSyncIdentityTable (
      id     INTEGER PRIMARY KEY CHECK (id = 1),
      origin TEXT NOT NULL,
      -- The env id of the last envelope WE deposited. Kept so the next deposit
      -- can retire the previous one: we must never acknowledge our own live
      -- envelope, because in a shared mailbox that is exactly the message the
      -- other device still has to read.
      last_deposit TEXT
    )
  ''');
  await db.execute('''
    CREATE TABLE IF NOT EXISTS $kSyncPeerStateTable (
      peer_uuid    TEXT PRIMARY KEY,
      -- How far THIS peer has confirmed applying of OUR rows. Only their echo
      -- moves it.
      cursor       INTEGER NOT NULL DEFAULT $kCursorUnsynced,
      -- How far WE have applied of THEIR rows. This is what we echo back, and
      -- keeping it separate is what stops the two directions from being
      -- conflated into one number that is wrong for both.
      applied_high INTEGER NOT NULL DEFAULT 0,
      updated_at   REAL NOT NULL
    )
  ''');
  // The relay guarantees at-least-once delivery, never exactly-once, so
  // idempotency has to live here: the same envelope arriving twice must be a
  // no-op, not a double apply.
  await db.execute('''
    CREATE TABLE IF NOT EXISTS $kSyncAppliedTable (
      env_id     TEXT PRIMARY KEY,
      applied_at REAL NOT NULL
    )
  ''');
  // Every revision that LOST a merge. The UNIQUE key is what stops a
  // redelivered envelope from making the user stare at the same decision twice.
  await db.execute('''
    CREATE TABLE IF NOT EXISTS $kSyncConflictsTable (
      id             INTEGER PRIMARY KEY AUTOINCREMENT,
      uuid           TEXT NOT NULL,
      losing_lamport INTEGER NOT NULL,
      losing_origin  TEXT,
      losing_payload TEXT NOT NULL,
      resolved_at    REAL NOT NULL,
      UNIQUE(uuid, losing_lamport, losing_origin, losing_payload)
    )
  ''');

  // Tables that already existed keep their OLD shape: `CREATE TABLE IF NOT
  // EXISTS` does nothing at all when the table is there, columns and all.
  //
  // Shipped without this, and the devices said so:
  //
  //   no such column: last_deposit
  //   UPDATE sync_identity SET last_deposit = ? WHERE id = 1
  //
  // Every install that had enabled sync before the change failed every pass,
  // while a FRESH install worked perfectly — which is precisely why the suite
  // stayed green: it only ever created new tables.
  // One row per DESTINATION mailbox: with a mailbox per device there is a
  // different "previous envelope" to retire for each peer, and a single column
  // could only ever remember the last one — retiring the wrong device's
  // message, or none.
  await db.execute('''
    CREATE TABLE IF NOT EXISTS $kSyncDepositsTable (
      mailbox TEXT PRIMARY KEY,
      env_id  TEXT NOT NULL
    )
  ''');

  await _addColumnIfMissing(db, kSyncIdentityTable, 'last_deposit', 'TEXT');
  await _addColumnIfMissing(
      db, kSyncPeerStateTable, 'applied_high', 'INTEGER NOT NULL DEFAULT 0');
}

/// ALTER, never re-CREATE.
///
/// Dropping and rebuilding `sync_identity` would hand the device a NEW origin,
/// and every row it had already authored would suddenly look like a stranger's
/// — turning its own history into a permanent conflict with itself.
Future<void> _addColumnIfMissing(
  DatabaseExecutor db,
  String table,
  String column,
  String definition,
) async {
  final info = await db.rawQuery('PRAGMA table_info($table)');
  final present = {for (final row in info) row['name'] as String?};
  if (present.contains(column)) return;
  await db.execute('ALTER TABLE $table ADD COLUMN $column $definition');
}

/// This device's identity inside the user's own device set.
///
/// Generated once and persisted. NEVER derived from the recovery phrase or any
/// hardware id: the phrase is shared by every device (so it could not tell them
/// apart) and a hardware id would follow the user across a factory reset and
/// leak into the tiebreak.
///
/// The relay is never told this value — it only ever sees a mailbox id.
Future<String> localOrigin(DatabaseExecutor db) async {
  await ensureSyncTables(db);
  final existing = await db.query(kSyncIdentityTable, where: 'id = 1');
  if (existing.isNotEmpty) return existing.first['origin']! as String;

  final origin = _uuid.v4();
  // INSERT OR IGNORE, not plain INSERT: two writes racing at first launch must
  // not crash, and whichever lands first is equally valid.
  await db.rawInsert(
    'INSERT OR IGNORE INTO $kSyncIdentityTable(id, origin) VALUES (1, ?)',
    [origin],
  );
  final settled = await db.query(kSyncIdentityTable, where: 'id = 1');
  return settled.first['origin']! as String;
}

/// The next logical clock value for a local write.
///
/// Reads the high-water mark from the graph itself rather than keeping a
/// counter in memory. Slightly more work per write, and worth it: a cached
/// counter and the table can disagree after a crash, a restore from backup, or
/// a second isolate writing — and every one of those makes new rows quietly
/// unsyncable by reusing a number the peer already has.
///
/// This also absorbs INCOMING rows: applying a peer's row at lamport 40 lifts
/// the mark, so the next local write is 41 and cannot collide with it.
Future<int> nextLamport(DatabaseExecutor db) async {
  final rows = await db.rawQuery('''
    SELECT MAX(high) AS high FROM (
      SELECT COALESCE(MAX(lamport), 0) AS high FROM nodes
      UNION ALL
      SELECT COALESCE(MAX(lamport), 0) AS high FROM edges
    )
  ''');
  final current = rows.first['high'] as int? ?? 0;
  return current + 1;
}

const String kSyncDepositsTable = 'sync_deposits';

/// The env id of the last envelope we left in [mailbox], if any.
Future<String?> lastDepositTo(DatabaseExecutor db, String mailbox) async {
  await ensureSyncTables(db);
  final rows = await db.query(kSyncDepositsTable,
      where: 'mailbox = ?', whereArgs: [mailbox]);
  return rows.isEmpty ? null : rows.first['env_id'] as String?;
}

Future<void> rememberDepositTo(
  DatabaseExecutor db,
  String mailbox,
  String envId,
) async {
  await ensureSyncTables(db);
  await db.insert(
    kSyncDepositsTable,
    {'mailbox': mailbox, 'env_id': envId},
    conflictAlgorithm: ConflictAlgorithm.replace,
  );
}

/// The env id of the last envelope this device deposited, if any.
Future<String?> lastDeposit(DatabaseExecutor db) async {
  await ensureSyncTables(db);
  final rows = await db.query(kSyncIdentityTable, where: 'id = 1');
  if (rows.isEmpty) return null;
  return rows.first['last_deposit'] as String?;
}

Future<void> rememberDeposit(DatabaseExecutor db, String envId) async {
  await localOrigin(db); // guarantees the row exists
  await db.update(kSyncIdentityTable, {'last_deposit': envId}, where: 'id = 1');
}

/// Record an envelope as applied. False when it had already been seen.
Future<bool> rememberApplied(DatabaseExecutor db, String envId) async {
  await ensureSyncTables(db);
  final changed = await db.rawInsert(
    'INSERT OR IGNORE INTO $kSyncAppliedTable(env_id, applied_at) VALUES (?, ?)',
    [envId, DateTime.now().millisecondsSinceEpoch / 1000.0],
  );
  return changed != 0;
}

/// Give the rows that predate the clock a place in it.
///
/// THE BUG THIS EXISTS FOR, measured on real devices: the phone showed many
/// memories, the laptop two, and both reported a healthy sync. They were
/// synced — of everything the cursor could see.
///
/// Every row written before stamping shipped sits at `lamport = 0`. The first
/// pass sends them (0 > -1), the peer applies them and echoes a high-water of
/// 0, the sender advances its cursor to 0 — and from that moment every
/// remaining lamport-0 row fails `lamport > cursor` and is excluded FOR EVER.
/// Part of the graph crosses, the rest silently never does, and neither device
/// has anything to report.
///
/// `axi/src/axi/sync/stamping.py` has had this since the first slice. This is
/// the port that was never written.
///
/// Idempotent, and safe to call at every startup: only rows still at 0 are
/// touched. A row that ARRIVED from another device keeps its author —
/// overwriting it would make this device claim authorship of everything it
/// ever received, and the deterministic tiebreak would stop meaning anything.
Future<int> backfillSyncStamps(DatabaseExecutor db) async {
  await ensureSyncTables(db);
  final origin = await localOrigin(db);
  var next = await nextLamport(db);
  var touched = 0;

  for (final table in const ['nodes', 'edges']) {
    final stale = await db.query(
      table,
      // origin_node too, or the update below reads null for EVERY row and
      // stamps this device as the author of memories another one wrote. The
      // test caught exactly that.
      columns: ['uuid', 'origin_node'],
      where: 'lamport IS NULL OR lamport = 0',
      orderBy: 'created_at ASC, uuid ASC',
    );
    for (final row in stale) {
      await db.update(
        table,
        {
          'lamport': next,
          // COALESCE in spirit: only fill an author that is missing.
          'origin_node': row['origin_node'] ?? origin,
        },
        where: 'uuid = ? AND (lamport IS NULL OR lamport = 0)',
        whereArgs: [row['uuid']],
      );
      next++;
      touched++;
    }
  }
  return touched;
}
